include/matx/transforms/solve/solve_cudss.h

```diff
@@ -0,0 +1,286 @@
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+/////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <cudss.h>
+
+#include <numeric>
+
+#include "matx/core/cache.h"
+#include "matx/core/sparse_tensor.h"
+#include "matx/core/tensor.h"
+
+namespace matx {
+
+namespace detail {
+
+/**
+ * Parameters needed to execute a cuDSS direct SOLVE.
+ */
+struct SolveCUDSSParams_t {
+  MatXDataType_t dtype;
+  MatXDataType_t ptype;
+  MatXDataType_t ctype;
+  int rank;
+  cudaStream_t stream;
+  index_t nse;
+  index_t m;
+  index_t n;
+  index_t k;
+  // Matrix handles in cuDSS are data specific (unlike e.g. cuBLAS
+  // where the same plan can be shared between different data buffers).
+  void *ptrA0;
+  void *ptrA1;
+  void *ptrA2;
+  void *ptrA3;
+  void *ptrA4;
+  void *ptrB;
+  void *ptrC;
+};
+
+template <typename TensorTypeC, typename TensorTypeA, typename TensorTypeB>
+class SolveCUDSSHandle_t {
+public:
+  using TA = typename TensorTypeA::value_type;
+  using TB = typename TensorTypeB::value_type;
+  using TC = typename TensorTypeC::value_type;
+
+  static constexpr int RANKA = TensorTypeC::Rank();
+  static constexpr int RANKB = TensorTypeC::Rank();
+  static constexpr int RANKC = TensorTypeC::Rank();
+
+  SolveCUDSSHandle_t(TensorTypeC &c, const TensorTypeA &a, const TensorTypeB &b,
+                     cudaStream_t stream) {
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_INTERNAL)
+
+    static_assert(RANKA == 2);
+    static_assert(RANKB == 2);
+    static_assert(RANKC == 2);
+
+    MATX_ASSERT(a.Size(RANKA - 1) == b.Size(RANKB - 2), matxInvalidSize);
+    MATX_ASSERT(c.Size(RANKC - 1) == b.Size(RANKB - 1), matxInvalidSize);
+    MATX_ASSERT(c.Size(RANKC - 2) == a.Size(RANKA - 2), matxInvalidSize);
+
+    params_ = GetSolveParams(c, a, b, stream);
+
+    [[maybe_unused]] cudssStatus_t ret = cudssCreate(&handle_);
+    MATX_ASSERT(ret == CUDSS_STATUS_SUCCESS, matxSolverError);
+
+    // Create cuDSS handle for sparse matrix A.
+    static_assert(is_sparse_tensor_v<TensorTypeA>);
+    MATX_ASSERT(TypeToInt<typename TensorTypeA::pos_type> ==
+                    TypeToInt<typename TensorTypeA::crd_type>,
+                matxNotSupported);
+    cudaDataType itp = MatXTypeToCudaType<typename TensorTypeA::crd_type>();
+    cudaDataType dta = MatXTypeToCudaType<TA>();
+    cudssMatrixType_t mtp = CUDSS_MTYPE_GENERAL;
+    cudssMatrixViewType_t mvw = CUDSS_MVIEW_FULL;
+    cudssIndexBase_t bas = CUDSS_BASE_ZERO;
+    if constexpr (TensorTypeA::Format::isCSR()) {
+      ret = cudssMatrixCreateCsr(&matA_, params_.m, params_.k, params_.nse,
+                                 /*rowStart=*/params_.ptrA2,
+                                 /*rowEnd=*/nullptr, params_.ptrA4,
+                                 params_.ptrA0, itp, dta, mtp, mvw, bas);
+    } else {
+      MATX_THROW(matxNotSupported, "cuDSS currently only supports CSR");
```

Comment:
Should we eventually convert automatically?

Reply:
It would be nice if we could somehow cache the handles individually. But between libs, the handles are not the same (e.g. cuSparse and cuDSS CSR handles are slightly different)

Reply:
Sorry, I meant if the user calls `solve()` should we restrict it to CSR or should we do an async transform into a different type?

Reply:
Next, I plan to add a convert (dense-sparse/sparse-sparse) also. Perhaps we should (for now) just rely on the user to do such conversions explicitly [they can be expensive]